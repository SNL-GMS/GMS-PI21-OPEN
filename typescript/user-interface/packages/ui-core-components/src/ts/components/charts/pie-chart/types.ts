export interface PieChartProps {
  style: PieChartStyle;
  percent: number;
  className?: string;
  pieSliceClass?: string;
  status?: string;
}

export interface PieChartStyle {
  diameterPx: number;
  borderPx?: number;
}
