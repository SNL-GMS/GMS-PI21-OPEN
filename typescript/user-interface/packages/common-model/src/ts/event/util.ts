import type { Event, EventHypothesis, LocationSolution, PreferredEventHypothesis } from './types';

/**
 * Finds preferred event hypothesis for a given event and stage
 * If no preferred hypothesis exists for stage, will return most recent preferred hypothesis by stage
 *
 * @param event
 * @param openIntervalName
 * @param stageNames
 */
export const findPreferredEventHypothesis = (
  event: Event,
  openIntervalName: string,
  stageNames: string[]
): EventHypothesis => {
  const { preferredEventHypothesisByStage, eventHypotheses } = event;
  let preferredEventHypothesis: PreferredEventHypothesis;
  // We want display the latest preferredEventHypothesis need to use the stageNames order
  // from the workflowQuery so that if the preferredEventHypothesisByStage is not present, will
  // try and grab the previous stage in the list e.x open 'AL1' but preferredEventHypothesisByStage only has
  // 'Auto Network' so will end up using 'Auto Network'
  const index = stageNames.indexOf(openIntervalName);
  for (let i = index; i > -1; i -= 1) {
    preferredEventHypothesis = preferredEventHypothesisByStage.find(
      hypothesis => hypothesis.stage.name === stageNames[i]
    );

    if (preferredEventHypothesis) {
      break;
    }
  }

  if (!preferredEventHypothesis) {
    return undefined;
  }

  return eventHypotheses.find(
    hypothesis => hypothesis.id.hypothesisId === preferredEventHypothesis.preferred.id.hypothesisId
  );
};

/**
 * Finds the last non rejected parent event hypothesis for a given event and hypothesis
 * if no valid hypothesis is found it returns the first hypothesis
 *
 * @param event
 * @param rejectedHypothesis
 */
export const findEventHypothesisParent = (
  event: Event,
  rejectedHypothesis: EventHypothesis
): EventHypothesis => {
  // Loop backwards until we find a non-rejected hypothesis
  for (let i = rejectedHypothesis.parentEventHypotheses.length - 1; i >= 0; i -= 1) {
    const parentEventHypothesis = event.eventHypotheses.find(
      hypo => hypo.id.hypothesisId === rejectedHypothesis.parentEventHypotheses[i].id.hypothesisId
    );
    if (parentEventHypothesis && !parentEventHypothesis.rejected) {
      return parentEventHypothesis;
    }
  }
  return rejectedHypothesis[0];
};

/**
 * Finds the preferred location solution for a hypothesis falling back to the parents if the hypothesis is rejected
 *
 * @param eventHypothesisId hypothesis id to find the solution for
 * @param eventHypotheses list of hypotheses
 * @returns a location solution
 */
export const findPreferredLocationSolution = (
  eventHypothesisId: string,
  eventHypotheses: EventHypothesis[]
): LocationSolution => {
  const eventHypothesis = eventHypotheses.find(
    hypothesis => hypothesis.id.hypothesisId === eventHypothesisId
  );
  if (eventHypothesis.preferredLocationSolution) {
    return eventHypothesis.locationSolutions.find(
      ls => ls.id === eventHypothesis.preferredLocationSolution.id
    );
  }

  if (eventHypothesis.parentEventHypotheses) {
    const parentHypothesisId =
      eventHypothesis.parentEventHypotheses[eventHypothesis.parentEventHypotheses.length - 1].id
        .hypothesisId;

    const parentEventHypothesis = eventHypotheses.find(
      hypothesis => hypothesis.id.hypothesisId === parentHypothesisId
    );

    return parentEventHypothesis.locationSolutions.find(
      ls => ls.id === parentEventHypothesis.preferredLocationSolution.id
    );
  }

  return undefined;
};
