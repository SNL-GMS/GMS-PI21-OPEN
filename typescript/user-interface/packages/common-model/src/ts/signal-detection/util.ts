import type {
  ArrivalTimeMeasurementValue,
  FeatureMeasurement,
  SignalDetection,
  SignalDetectionHypothesis
} from './types';
import { FeatureMeasurementType } from './types';

/**
 * Get the current Hypothesis from the set of Hypotheses. This will be the last entry in the set if there is one
 * Returns undefined on empty arrays
 *
 * @param hypotheses the set of Hypotheses.
 * @return the current Hypothesis
 */
export function getCurrentHypothesis(
  hypotheses: SignalDetectionHypothesis[]
): SignalDetectionHypothesis {
  return hypotheses?.slice(-1)[0];
}

/**
 * Searches Feature Measurements for the desired Feature Measurement
 *
 * @param featureMeasurements List of feature measurements
 * @param featureMeasurementType Enum of desired Feature Measurement desired
 *
 * @returns FeatureMeasurement or undefined if not found
 */
export function findFeatureMeasurementByType(
  featureMeasurements: FeatureMeasurement[],
  featureMeasurementType: FeatureMeasurementType
): FeatureMeasurement | undefined {
  if (featureMeasurements && featureMeasurementType) {
    return featureMeasurements.find(fm => fm?.featureMeasurementType === featureMeasurementType);
  }
  return undefined;
}

/**
 * Searches a SignalDetection for the ArrivalTime Feature Measurement
 *
 * @param signalDetection to search for ArrivalTime
 *
 * @returns ArrivalTime FeatureMeasurement or undefined if not found
 */
export function findArrivalTimeFeatureMeasurementUsingSignalDetection(
  signalDetection: SignalDetection
): FeatureMeasurement | undefined {
  const currentHypo = getCurrentHypothesis(signalDetection.signalDetectionHypotheses);
  if (!currentHypo) {
    return undefined;
  }
  return findFeatureMeasurementByType(
    currentHypo.featureMeasurements,
    FeatureMeasurementType.ARRIVAL_TIME
  );
}

/**
 * Searches Feature Measurements for the ArrivalTime Feature Measurement
 *
 * @param featureMeasurements List of feature measurements
 *
 * @returns ArrivalTime FeatureMeasurement or undefined if not found
 */
export function findArrivalTimeFeatureMeasurement(
  featureMeasurements: FeatureMeasurement[]
): FeatureMeasurement | undefined {
  return findFeatureMeasurementByType(featureMeasurements, FeatureMeasurementType.ARRIVAL_TIME);
}

/**
 * Searches Feature Measurements for the ArrivalTime Feature Measurement Value
 *
 * @param featureMeasurements List of feature measurements
 *
 * @returns ArrivalTime FeatureMeasurementValue or undefined if not found
 */
export function findArrivalTimeFeatureMeasurementValue(
  featureMeasurements: FeatureMeasurement[]
): ArrivalTimeMeasurementValue | undefined {
  const fm = findArrivalTimeFeatureMeasurement(featureMeasurements);
  return fm ? (fm.measurementValue as ArrivalTimeMeasurementValue) : undefined;
}
