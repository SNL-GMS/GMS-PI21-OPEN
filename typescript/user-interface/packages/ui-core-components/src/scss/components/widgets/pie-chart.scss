.pie-chart {
  &__pie-slice {
    fill: transparent;
    transform-origin: 50% 50%;
    transform: rotate(-90deg);
    stroke: var(--core-main, var(--gms-main, $core-main)); // default
  }
}
