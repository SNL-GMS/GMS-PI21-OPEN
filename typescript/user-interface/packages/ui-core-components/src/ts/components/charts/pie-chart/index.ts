import { PieChart } from './pie-chart';
import type { PieChartProps } from './types';

export { PieChart, PieChartProps };
