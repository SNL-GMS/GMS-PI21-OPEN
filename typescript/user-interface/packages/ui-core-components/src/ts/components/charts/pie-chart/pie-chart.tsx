/* eslint-disable react/prop-types */
import { getDataAttributesFromProps } from '@gms/ui-util';
import * as React from 'react';

import type { PieChartProps, PieChartStyle } from './types';

const PIE_CHART = 'pie-chart';
const SLICE = 'pie-slice';
const PIE_CHART_SLICE = `${PIE_CHART}__${SLICE}`;

const defaultBorderWidth = 2;

const getTotalPieChartDiameter = (radiusPx: number, borderPx: number) =>
  radiusPx * 2 + borderPx * 2;

const getRadiusWithBorder = (radiusPx: number, style: PieChartStyle) =>
  radiusPx + (style.borderPx ?? defaultBorderWidth);

// eslint-disable-next-line react/function-component-definition
const PieSlice: React.FunctionComponent<PieChartProps> = ({
  style,
  percent,
  pieSliceClass,
  ...rest
}) => {
  const pieRadiusPx = style.diameterPx / 2; // an inner circle at r/2
  const circumferenceOfInnerCirclePx = pieRadiusPx * Math.PI; // an inner circle at r/2
  const dataAttributes = getDataAttributesFromProps(rest);
  return (
    <circle
      className={`${PIE_CHART_SLICE} ${pieSliceClass ?? ''}`}
      r={pieRadiusPx / 2}
      cx={pieRadiusPx}
      cy={pieRadiusPx}
      strokeWidth={pieRadiusPx}
      strokeDasharray={`${circumferenceOfInnerCirclePx * percent} ${circumferenceOfInnerCirclePx} `}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...dataAttributes}
    />
  );
};

/**
 *
 * @param diameterPx the diameter, in px, of the indicator background
 * @param percent the percent, from 0 to 1.
 */
// eslint-disable-next-line react/function-component-definition
const CorePieChart: React.FunctionComponent<PieChartProps> = ({
  style,
  percent,
  className,
  pieSliceClass,
  ...rest
}) => {
  const radiusPx = style.diameterPx / 2;
  const borderRadiusPx = style.borderPx ?? defaultBorderWidth;
  const totalDiameterPx = getTotalPieChartDiameter(radiusPx, borderRadiusPx);
  const dataAttributes = getDataAttributesFromProps(rest);
  return (
    <svg
      className={`${PIE_CHART} ${className}`}
      width={`${totalDiameterPx}px`}
      height={`${totalDiameterPx}px`}
    >
      <g className={`${PIE_CHART}__container`}>
        <circle
          className={`
            ${PIE_CHART}
            ${className}`}
          r={getRadiusWithBorder(radiusPx, style)}
          cx={radiusPx}
          cy={radiusPx}
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...dataAttributes}
        />
        <PieSlice
          style={style}
          percent={percent}
          pieSliceClass={pieSliceClass}
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...dataAttributes}
        />
      </g>
    </svg>
  );
};

export const PieChart = React.memo(CorePieChart);
