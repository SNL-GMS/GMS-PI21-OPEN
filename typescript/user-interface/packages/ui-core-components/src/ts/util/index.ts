export * from './display-util';
export * from './non-ideal-state';
