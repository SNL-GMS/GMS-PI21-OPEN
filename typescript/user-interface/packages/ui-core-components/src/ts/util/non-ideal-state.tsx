import type { IconName } from '@blueprintjs/core';
import { Intent, NonIdealState, Spinner } from '@blueprintjs/core';
import React from 'react';

/**
 * Creates a NonIdealState blueprint JSX with the option of passing a title and description
 *
 * @param title optional description of the title that will be displayed
 * @param description optional description that will be displayed
 * @param icon optional the icon state that will be displayed
 */
export function nonIdealStateWithNoSpinner(
  title?: string,
  description?: string,
  icon?: IconName
): any {
  return <NonIdealState icon={icon} title={title || null} description={description || null} />;
}

/**
 * Creates a NonIdealState blueprint JSX with a spinner action
 * with the option of passing a title and description
 *
 * @param title optional description of the title that will be displayed
 * @param description optional description that will be displayed
 * @param icon optional the icon state that will be displayed
 * @param intent optional the intent state that will be displayed
 */
export function nonIdealStateWithSpinner(
  title?: string,
  description?: string,
  icon?: IconName,
  intent: Intent = Intent.PRIMARY
): JSX.Element {
  return (
    <NonIdealState
      action={<Spinner intent={intent} />}
      icon={icon}
      title={title || null}
      description={description || null}
    />
  );
}
