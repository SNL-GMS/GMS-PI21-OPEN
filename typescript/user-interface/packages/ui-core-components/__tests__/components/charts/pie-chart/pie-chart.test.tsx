/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-magic-numbers */
import { render } from '@testing-library/react';
import * as Enzyme from 'enzyme';
import * as React from 'react';

import { PieChart } from '../../../../src/ts/components/charts/pie-chart/pie-chart';
import type * as PieChartTypes from '../../../../src/ts/components/charts/pie-chart/types';
import { waitForComponentToPaint } from '../../../util/test-util';

const pieChartStyle: PieChartTypes.PieChartStyle = {
  diameterPx: 100,
  borderPx: 5
};

const props: PieChartTypes.PieChartProps = {
  style: pieChartStyle,
  percent: 40,
  className: 'string',
  pieSliceClass: 'string',
  status: 'string'
};

describe('Pie Chart', () => {
  it('to be defined', () => {
    expect(PieChart).toBeDefined();
  });

  it('Pie Chart renders', () => {
    const { container } = render(<PieChart {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('History List functions and clicks', async () => {
    const wrapper3 = Enzyme.mount(<PieChart {...props} />);

    wrapper3.setProps({ percent: 50 });
    await waitForComponentToPaint(wrapper3);
    expect(wrapper3.find('PieSlice')).toBeDefined();
  });
});
