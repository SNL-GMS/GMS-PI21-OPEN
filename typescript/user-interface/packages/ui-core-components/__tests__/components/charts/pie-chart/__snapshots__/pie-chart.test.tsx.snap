// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pie Chart Pie Chart renders 1`] = `
<div>
  <svg
    class="pie-chart string"
    height="110px"
    width="110px"
  >
    <g
      class="pie-chart__container"
    >
      <circle
        class="
            pie-chart
            string"
        cx="50"
        cy="50"
        r="55"
      />
      <circle
        class="pie-chart__pie-slice string"
        cx="50"
        cy="50"
        r="25"
        stroke-dasharray="6283.185307179587 157.07963267948966 "
        stroke-width="50"
      />
    </g>
  </svg>
</div>
`;
